aibolit/patterns/mutable_index/mutable_index.py

```diff
@@ -2,21 +2,66 @@
 # SPDX-License-Identifier: MIT
 import os
 
+from aibolit.ast_framework.ast import AST
+from aibolit.ast_framework.ast_node import ASTNode
+from aibolit.ast_framework.ast_node_type import ASTNodeType
+from aibolit.types_decl import LineNumber
+from aibolit.utils.ast_builder import build_ast
 
-class MutableIndex:
 
+class MutableIndex:
     def __init__(self):
         pass
 
-    def value(self, filename: str | os.PathLike):
+    def value(self, filename: str | os.PathLike) -> list[LineNumber]:
         """
         Traverse over AST tree and finds loops with mutable index
 
         :return: List of line number of loops with mutable index
-
-        @todo #147:30min Implement MutableIndex pattern
-        We need to implement the code on mutable index pattern. It must return the number
-        of the lines where we have an index of a for loop being mutated. After implementing,
-        the method, enable all tests in test_mutable_index.py
         """
-        return []
+        result = set()
+        ast = AST.build_from_javalang(build_ast(filename))
+        for node in ast.get_proxy_nodes(
+            ASTNodeType.FOR_STATEMENT,
+        ):
+            index_names = self._collect_index_names(ast, node)
+            for_body = ast.get_subtree(node.body)
+            result.update(self._handle_assignment(index_names, for_body))
+            result.update(self._handle_unary_operations(index_names, for_body))
+
+        return sorted(result)
+
+    def _collect_index_names(self, ast: AST, node: ASTNode) -> set[str]:
+        result = set()
+        node_control = ast.get_subtree(node.control)
+        for decl in node_control.get_proxy_nodes(ASTNodeType.VARIABLE_DECLARATOR):
+            result.add(decl.name)
+        for assignment in node_control.get_proxy_nodes(ASTNodeType.ASSIGNMENT):
+            result.add(assignment.expressionl.member)
+        return result
+
+    def _handle_assignment(self, index_names: set[str], for_body: AST) -> set[LineNumber]:
+        result = set()
+        for op in for_body.get_proxy_nodes(ASTNodeType.ASSIGNMENT):
+            if op.expressionl.node_type != ASTNodeType.MEMBER_REFERENCE:
+                continue
+            if op.expressionl.member not in index_names:
+                continue
+            result.add(op.expressionl.line)
```

Comment:
@AntonProkopyev as a way to reduce cognitive complexity (avoid `continue`, if possible/reasonable, and use positive conditions in `if` statement) can we refactor it to something like this?

```suggestion
           if (
               op.expressionl.node_type == ASTNodeType.MEMBER_REFERENCE and
                op.expressionl.member in index_names
           ):
               result.add(op.expressionl.line)
```